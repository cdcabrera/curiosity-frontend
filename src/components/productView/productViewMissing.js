import React from 'react';
import PropTypes from 'prop-types';
import { Button, Card, CardBody, CardFooter, CardTitle, Gallery, Title, PageSection } from '@patternfly/react-core';
import { ArrowRightIcon } from '@patternfly/react-icons';
import { useMount } from 'react-use';
import { PageLayout, PageHeader } from '../pageLayout/pageLayout';
import { useRouteDetail, useNavigate } from '../router/routerContext';
import { helpers } from '../../common';
import { translate } from '../i18n/i18n';

/**
 * Render a missing product view.
 *
 * @fires onNavigate
 * @param {object} props
 * @param {number} props.availableProductsRedirect
 * @param {Function} props.t
 * @param {Function} props.useNavigate
 * @param {Function} props.useRouteDetail
 * @returns {Node}
 */
const ProductViewMissing = ({
  availableProductsRedirect,
  t,
  useNavigate: useAliasNavigate,
  useRouteDetail: useAliasRouteDetail
}) => {
  const navigate = useAliasNavigate();
  const { productConfig, allProductConfigs } = useAliasRouteDetail();
  const availableProducts = (productConfig?.length && productConfig) || allProductConfigs;

  useMount(() => {
    if (availableProducts.length <= availableProductsRedirect) {
      navigate(availableProducts[0].productPath);
    }
  });

  /**
   * On click, update history.
   *
   * @event onNavigate
   * @param {string} id
   * @returns {void}
   */
  const onNavigate = id => navigate(id);

  return (
    <PageLayout className="curiosity-missing-view">
      <PageHeader productLabel="missing">{t(`curiosity-view.title`, { appName: helpers.UI_DISPLAY_NAME })}</PageHeader>
      <PageSection isFilled>
        <Gallery hasGutter>
          {availableProducts.map(({ productGroup, productId }) => (
            <Card key={`missingViewCard-${productId}`} isHoverable onClick={() => onNavigate(productId)}>
              <CardTitle>
                <Title headingLevel="h2" size="lg">
                  {t('curiosity-view.title', {
                    appName: helpers.UI_DISPLAY_NAME,
                    context: (Array.isArray(productId) && productId?.[0]) || productId
                  })}
                </Title>
              </CardTitle>
              <CardBody className="curiosity-missing-view__card-description">
                {t('curiosity-view.description', {
                  appName: helpers.UI_DISPLAY_NAME,
                  context: (Array.isArray(productGroup) && productGroup?.[0]) || productGroup
                })}
              </CardBody>
              <CardFooter>
                <Button
                  variant="link"
                  isInline
                  onClick={event => {
                    event.preventDefault();
                    onNavigate(productId);
                  }}
                  icon={<ArrowRightIcon />}
                  iconPosition="right"
                >
                  Open
                </Button>
              </CardFooter>
            </Card>
          ))}
        </Gallery>
      </PageSection>
    </PageLayout>
  );
};

/**
 * Prop types.
 *
 * @type {{useNavigate: Function, availableProductsRedirect: number, t: Function, useRouteDetail: Function}}
 */
ProductViewMissing.propTypes = {
  availableProductsRedirect: PropTypes.number,
  t: PropTypes.func,
  useNavigate: PropTypes.func,
  useRouteDetail: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useNavigate: Function, availableProductsRedirect: number, t: translate, useRouteDetail: Function}}
 */
ProductViewMissing.defaultProps = {
  availableProductsRedirect: 4,
  t: translate,
  useNavigate,
  useRouteDetail
};

export { ProductViewMissing as default, ProductViewMissing };
