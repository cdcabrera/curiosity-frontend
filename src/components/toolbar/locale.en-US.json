{
  "curiosity-toolbar": {
    "button_displayName": "Search button for name",
    "clearFilters": "Reset filters",
    "label": "",
    "label_billing_provider": "",
    "label_billing_provider_none": "None",
    "label_billing_provider_aws": "Amazon Web Services",
    "label_billing_provider_azure": "Microsoft Azure",
    "label_billing_provider_gcp": "Google Cloud Platform",
    "label_billing_provider_oracle": "Oracle Cloud Infrastructure",
    "label_billing_provider_red hat": "Red Hat Marketplace",
    "label_category_cloud": "Public cloud",
    "label_category_hypervisor": "Hypervisor",
    "label_category_physical": "Physical",
    "label_category_virtual": "Virtual",
    "label_filter": "Filter by",
    "label_filter_billing_provider": "Purchased through",
    "label_filter_category": "Type",
    "label_filter_granularity": "Granularity",
    "label_filter_rangedMonthly": "Monthly",
    "label_filter_sla": "SLA",
    "label_filter_uom": "Unit of measure",
    "label_filter_usage": "Usage",
    "label_filter_variant": "Variant",
    "label_granularity": "",
    "label_granularity_Daily": "Past 30 days (daily)",
    "label_granularity_Weekly": "Past 12 weeks (weekly)",
    "label_granularity_Monthly": "Past year (monthly)",
    "label_granularity_Quarterly": "Past 3 years (quarterly)",
    "label_granularityRangedMonthly": "{{context}}",
    "label_granularityRangedMonthly_current": "This month",
    "label_groupVariant": "Variant:",
    "label_groupVariant_OpenShift Container Platform": "Container Platform (Annual)",
    "label_groupVariant_OpenShift-dedicated-metrics": "Dedicated (On-Demand)",
    "label_groupVariant_OpenShift-metrics": "Container Platform (On-Demand)",
    "label_groupVariant_RHEL for ARM": "RHEL ARM",
    "label_groupVariant_RHEL for IBM Power": "RHEL for IBM Power",
    "label_groupVariant_RHEL for IBM z": "RHEL for IBM z",
    "label_groupVariant_RHEL for x86": "RHEL for x86",
    "label_groupVariant_rhel-for-x86-eus": "RHEL for x86 EUS",
    "label_groupVariant_rhel-for-sap-x86": "RHEL for SAP x86",
    "label_groupVariant_rosa": "ROSA Hosted Control Planes",
    "label_groupVariant_Satellite Capsule": "Satellite Capsule",
    "label_groupVariant_Satellite Server": "Satellite Server",
    "label_sla": "",
    "label_sla_none": "No SLA",
    "label_sla_Premium": "Premium",
    "label_sla_Self-Support": "Self-Support",
    "label_sla_Standard": "Standard",
    "label_uom": "",
    "label_uom_none": "None",
    "label_uom_cores": "Cores",
    "label_uom_sockets": "Sockets",
    "label_usage": "",
    "label_usage_none": "Unspecified",
    "label_usage_Development/Test": "Development/Test",
    "label_usage_Disaster Recovery": "Disaster Recovery",
    "label_usage_Production": "Production",
    "placeholder": "Select",
    "placeholder_billing_provider": "Select purchased through",
    "placeholder_granularity": "Select date range",
    "placeholder_groupVariant": "Select a variant",
    "placeholder_rangedMonthly": "Select a month",
    "placeholder_sla": "Select SLA",
    "placeholder_uom": "Select unit",
    "placeholder_usage": "Select usage",
    "placeholder_filter": "Filter by",
    "placeholder_filter_billing_provider": "Filter by purchased through",
    "placeholder_filter_category": "Filter by type",
    "placeholder_filter_displayName": "Filter by name",
    "placeholder_filter_granularity": "Filter by date",
    "placeholder_filter_groupVariant": "Filter by variant",
    "placeholder_filter_rangedMonthly": "Filter by month",
    "placeholder_filter_sla": "Filter by SLA",
    "placeholder_filter_uom": "Filter by unit",
    "placeholder_filter_usage": "Filter by usage",
    "placeholder_filter_variant": "Filter by variant"
  }
}
