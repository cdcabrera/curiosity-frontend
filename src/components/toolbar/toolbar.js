import React from 'react';
import PropTypes from 'prop-types';
import {
  Toolbar as PfToolbar,
  ToolbarContent,
  ToolbarFilter,
  ToolbarGroup,
  ToolbarItem,
  ToolbarToggleGroup
} from '@patternfly/react-core';
import { FilterIcon } from '@patternfly/react-icons';
import _isEqual from 'lodash/isEqual';
import { connect, reduxTypes, store, useSelector } from '../../redux';
import { RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { helpers } from '../../common';
import { translate } from '../i18n/i18n';
import { ToolbarFieldSla, toolbarFieldOptions as slaOptions } from './toolbarFieldSla';
import { ToolbarFieldUsage, toolbarFieldOptions as usageOptions } from './toolbarFieldUsage';
import { ToolbarFieldCategory, toolbarFieldOptions as categoryOptions } from './toolbarFieldCategory';

/**
 * Application filter toolbar.
 *
 * @augments React.Component
 * @fires onClearAllFilters
 * @fires onClearFilter
 */
/*
class Toolbar extends React.Component {
  componentDidUpdate(prevProps) {
    const { query } = this.props;
    console.log('onupdate >>>', _isEqual(query, prevProps.query));
    // if (!_isEqual(query, prevProps.query)) {
    // }
  }

  /**
   * Clear all filters' state.
   *
   * @event onClearAllFilters
   * /
  onClearAllFilters = () => {
    const { viewId } = this.props;

    store.dispatch({
      type: reduxTypes.query.SET_QUERY_CLEAR,
      viewId,
      clearFilters: {
        [RHSM_API_QUERY_TYPES.SLA]: null,
        [RHSM_API_QUERY_TYPES.USAGE]: null
      }
    });
  };

  /**
   * Clear individual filter state.
   *
   * @event onClearFilter
   * @param {string} categoryTitle
   * /
  onClearFilter = categoryTitle => {
    const { viewId } = this.props;
    const { value: categoryValue } = categoryOptions.find(({ title }) => title === categoryTitle) || {};

    if (!categoryValue) {
      return;
    }

    store.dispatch({
      type: reduxTypes.query.SET_QUERY_CLEAR,
      viewId,
      clearFilters: {
        [categoryValue]: null
      }
    });
  };

  /**
   * Return the current category filter, fallback to props selected.
   *
   * @returns {string|undefined}
   * /
  getCurrentCategoryFilter() {
    const { currentFilter, filterOptions } = this.props;

    return (
      currentFilter ||
      filterOptions.find(({ selected }) => selected === true)?.id ||
      (filterOptions.length === 1 && filterOptions[0]?.id)
    );
  }

  /**
   * Available, and selected select filter options for chip display.
   *
   * @param {string} field
   * @param {Array} options
   * @returns {Array}
   * /
  getSelectedFilterForChips(field, options) {
    const { query } = this.props;
    const filter = typeof query?.[field] === 'string' && options.find(({ value }) => value === query?.[field]);

    return (filter?.title && [filter.title]) || [];
  }

  /**
   * Render a filter node.
   *
   * @param {string} field
   * @returns {Node}
   * /
  renderFiltersWORKSISH(field) {
    const { t, viewId } = this.props;
    const currentCategoryFilter = this.getCurrentCategoryFilter();
    let selectedFilterForChips;
    let filter;

    switch (field) {
      case RHSM_API_QUERY_TYPES.USAGE:
        selectedFilterForChips = this.getSelectedFilterForChips(field, usageOptions);
        filter = <ToolbarFieldUsage viewId={viewId} />;
        break;
      case RHSM_API_QUERY_TYPES.SLA:
      default:
        selectedFilterForChips = this.getSelectedFilterForChips(field, slaOptions);
        filter = <ToolbarFieldSla viewId={viewId} />;
        break;
    }

    return (
      <ToolbarFilter
        key={field}
        chips={selectedFilterForChips}
        deleteChip={this.onClearFilter}
        categoryName={t('curiosity-toolbar.category', { context: field })}
        showToolbarItem={currentCategoryFilter === field}
      >
        {filter}
      </ToolbarFilter>
    );
  }

  renderFilters() {
    const { t, viewId } = this.props;
    const currentCategoryFilter = this.getCurrentCategoryFilter();

    return (
      <React.Fragment>
        <ToolbarFilter
          key={RHSM_API_QUERY_TYPES.SLA}
          chips={this.getSelectedFilterForChips(RHSM_API_QUERY_TYPES.SLA, slaOptions)}
          deleteChip={this.onClearFilter}
          categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.SLA })}
          // showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.SLA}
        >
          <ToolbarFieldSla viewId={viewId} />
        </ToolbarFilter>
        <ToolbarFilter
          key={RHSM_API_QUERY_TYPES.USAGE}
          chips={this.getSelectedFilterForChips(RHSM_API_QUERY_TYPES.USAGE, usageOptions)}
          deleteChip={this.onClearFilter}
          categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.USAGE })}
          // showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.USAGE}
        >
          <ToolbarFieldUsage viewId={viewId} />
        </ToolbarFilter>
      </React.Fragment>
    );
  }

  /**
   * Render toolbar category select.
   *
   * @returns {Node}
   * /
  renderCategories() {
    const { filterOptions, viewId } = this.props;
    const currentCategoryFilter = this.getCurrentCategoryFilter();
    const options = categoryOptions.filter(({ value }) => filterOptions.find(({ id }) => id === value));

    if (options.length > 1) {
      return (
        <ToolbarItem>
          <ToolbarFieldCategory
            viewId={viewId}
            options={options}
            value={currentCategoryFilter}
            toggleIcon={<FilterIcon />}
          />
        </ToolbarItem>
      );
    }

    return null;
  }

  /**
   * Render a filter toolbar.
   *
   * @returns {Node}
   * /
  render() {
    const { chipGroupSequence, filterOptions, isDisabled } = this.props;

    if (isDisabled) {
      return null;
    }

    // let updatedFilterOptions = filterOptions;

    /*
    if (chipGroupSequence.size) {
      updatedFilterOptions = [];
      Array.from(chipGroupSequence).forEach(field => {
        const filter = filterOptions.find(obj => obj.id === field);
        updatedFilterOptions.push(filter);
      });
    }
    * /

    return (
      <PfToolbar
        id="curiosity-toolbar"
        className="curiosity-toolbar pf-m-toggle-group-container ins-c-primary-toolbar"
        collapseListedFiltersBreakpoint="sm"
        clearAllFilters={this.onClearAllFilters}
      >
        <ToolbarContent>
          <ToolbarToggleGroup toggleIcon={<FilterIcon />} breakpoint="md">
            <ToolbarGroup variant="filter-group">
              {this.renderCategories()}
              {
                this.renderFilters()
                // updatedFilterOptions.map(({ id }) => this.renderFilters(id))
                // .sort(({ id: a }, { id: b }) => a && a.localeCompare(b))}
              }
            </ToolbarGroup>
          </ToolbarToggleGroup>
        </ToolbarContent>
      </PfToolbar>
    );
  }
}
*/

const Toolbar = ({ filterOptions, isDisabled, query, t, viewId }) => {
  const currentFilter = useSelector(({ toolbar }) => toolbar.filters?.[viewId]?.currentFilter, null);
  const updatedQuery = useSelector(({ view }) => view.query?.[viewId], query);

  if (isDisabled) {
    return null;
  }

  const updatedFilterOptions = categoryOptions.filter(({ value }) => filterOptions.find(({ id }) => id === value));
  const currentCategoryFilter =
    currentFilter ||
    filterOptions.find(({ selected }) => selected === true)?.id ||
    (filterOptions.length === 1 && filterOptions[0]?.id);

  const getSelectedFilterForChips = (field, options) => {
    const filter =
      typeof updatedQuery?.[field] === 'string' && options.find(({ value }) => value === updatedQuery?.[field]);
    return (filter?.title && [filter.title]) || [];
  };

  const onClearAllFilters = () =>
    store.dispatch({
      type: reduxTypes.query.SET_QUERY_CLEAR,
      viewId,
      clearFilters: {
        [RHSM_API_QUERY_TYPES.SLA]: null,
        [RHSM_API_QUERY_TYPES.USAGE]: null
      }
    });

  const onClearFilter = categoryTitle => {
    const { value: categoryValue } = categoryOptions.find(({ title }) => title === categoryTitle) || {};
    if (categoryValue) {
      store.dispatch({
        type: reduxTypes.query.SET_QUERY_CLEAR,
        viewId,
        clearFilters: {
          [categoryValue]: null
        }
      });
    }
  };

  return (
    <PfToolbar
      id="curiosity-toolbar"
      className="curiosity-toolbar pf-m-toggle-group-container ins-c-primary-toolbar"
      collapseListedFiltersBreakpoint="sm"
      clearAllFilters={onClearAllFilters}
    >
      <ToolbarContent>
        <ToolbarToggleGroup toggleIcon={<FilterIcon />} breakpoint="md">
          <ToolbarGroup variant="filter-group">
            {updatedFilterOptions.length > 1 && (
              <ToolbarItem>
                <ToolbarFieldCategory
                  viewId={viewId}
                  options={updatedFilterOptions}
                  value={currentCategoryFilter}
                  toggleIcon={<FilterIcon />}
                />
              </ToolbarItem>
            )}
            <ToolbarFilter
              key={RHSM_API_QUERY_TYPES.SLA}
              chips={getSelectedFilterForChips(RHSM_API_QUERY_TYPES.SLA, slaOptions)}
              deleteChip={onClearFilter}
              categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.SLA })}
              showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.SLA}
            >
              <ToolbarFieldSla viewId={viewId} />
            </ToolbarFilter>
            <ToolbarFilter
              key={RHSM_API_QUERY_TYPES.USAGE}
              chips={getSelectedFilterForChips(RHSM_API_QUERY_TYPES.USAGE, usageOptions)}
              deleteChip={onClearFilter}
              categoryName={t('curiosity-toolbar.category', { context: RHSM_API_QUERY_TYPES.USAGE })}
              showToolbarItem={currentCategoryFilter === RHSM_API_QUERY_TYPES.USAGE}
            >
              <ToolbarFieldUsage viewId={viewId} />
            </ToolbarFilter>
          </ToolbarGroup>
        </ToolbarToggleGroup>
      </ToolbarContent>
    </PfToolbar>
  );
};

/**
 * Prop types
 *
 * @type {{viewId: string, t: Function, activeFilters: Set, hardFilterReset: boolean, query: object,
 *     currentFilter: string, isDisabled: boolean, Array}}
 */
Toolbar.propTypes = {
  query: PropTypes.shape({
    [RHSM_API_QUERY_TYPES.SLA]: PropTypes.string,
    [RHSM_API_QUERY_TYPES.USAGE]: PropTypes.string
  }),
  // chipGroupSequence: PropTypes.instanceOf(Set),
  // activeFilters: PropTypes.instanceOf(Set),
  // currentFilter: PropTypes.oneOf([RHSM_API_QUERY_TYPES.SLA, RHSM_API_QUERY_TYPES.USAGE]),
  filterOptions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOf([RHSM_API_QUERY_TYPES.SLA, RHSM_API_QUERY_TYPES.USAGE]),
      selected: PropTypes.bool
    })
  ),
  // hardFilterReset: PropTypes.bool,
  isDisabled: PropTypes.bool,
  t: PropTypes.func,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{viewId: string, t: translate, activeFilters: Set, hardFilterReset: boolean, query: object,
 *     currentFilter: string, isDisabled: boolean, filterOptions: Array}}
 */
Toolbar.defaultProps = {
  query: {},
  // activeFilters: new Set(),
  // chipGroupSequence: new Set(),
  // currentFilter: null,
  filterOptions: [
    {
      id: RHSM_API_QUERY_TYPES.SLA
    },
    {
      id: RHSM_API_QUERY_TYPES.USAGE,
      selected: true
    }
  ],
  // hardFilterReset: false,
  isDisabled: helpers.UI_DISABLED_TOOLBAR,
  t: translate,
  viewId: 'toolbar'
};

/**
 * Apply state to props.
 *
 * @param {object} state
 * @param {object} state.toolbar
 * @param {object} state.view
 * @param {object} props
 * @param {string} props.viewId
 * @returns {object}
 */
/*
const mapStateToProps = ({ toolbar, view }, { viewId }) => ({
  query: view?.query?.[viewId],
  ...toolbar.filters?.[viewId]
});

const ConnectedToolbar = connect(mapStateToProps)(Toolbar);

export { ConnectedToolbar as default, ConnectedToolbar, Toolbar };
 */
export { Toolbar as default, Toolbar };
