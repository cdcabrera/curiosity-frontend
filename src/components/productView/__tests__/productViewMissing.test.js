import React from 'react';
import { shallow, mount } from 'enzyme';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { Button, Card } from '@patternfly/react-core';
import ReactRouterDom, { useHistory, MemoryRouter } from 'react-router-dom';
import { ProductViewMissing } from '../productViewMissing';
import { Toolbar } from '../../toolbar/toolbar';
// import {act} from "react-dom/test-utils";

describe('ProductViewMissing Component', () => {
  let container = null;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });
  /*
  const loadHookComponent = async callback => {
    let component = null;
    await act(async () => {
      component = callback();
    });
    component?.update();
    return component;
  };
  */
  const loadHookComponent = callback => {
    let component = null;
    act(() => {
      component = callback();
    });
    component?.update();
    return component;
  };

  const mockWindowLocation = async ({ url, callback }) => {
    const updatedUrl = new URL(url);
    const { location } = window;
    delete window.location;
    // mock
    window.location = {
      href: updatedUrl.href,
      search: updatedUrl.search,
      hash: updatedUrl.hash,
      pathname: updatedUrl.pathname
    };
    // await loadHookComponent(callback);
    await callback();
    // restore
    window.location = location;
  };

  it('should render a non-connected component', () => {
    const props = {};
    const component = shallow(<ProductViewMissing {...props} />);
    expect(component).toMatchSnapshot('non-connected');
  });

  it('should render a predictable set of product cards', () => {
    const props = {};

    mockWindowLocation({
      url: 'https://ci.foo.redhat.com/loremIpsum/dolorSit/',
      callback: () => {
        const component = shallow(<ProductViewMissing {...props} />);
        expect(component).toMatchSnapshot('non-connected');
      }
    });
  });

  it('should have cards that navigate to product IDs when clicked, WORKSISH', () => {
    const mockPush = jest.fn();
    const props = {
      useHistory: () => ({
        push: mockPush
      })
    };

    mockWindowLocation({
      url: 'https://ci.foo.redhat.com/loremIpsum/dolorSit/',
      callback: () => {
        /*
        const mockPush = jest.fn();
        jest.mock('react-router-dom', () => ({
          ...jest.requireActual('react-router-dom'),
          useHistory: () => ({
            push: mockPush
          })
        }));
        */
        // const mockPush = jest.fn();
        /*
        jest.mock('react-router-dom', () => {
          const Test = function () { // eslint-disable-line
            this.useHistory = () => this;
            this.push = () => mockPush;
          };
          return new Test();
        });
        */

        const component = shallow(<ProductViewMissing {...props} />);
        const cardButton = component.find(Card).first().find(Button).first();

        cardButton.simulate('click', 'lorem-ipsum-id');

        // console.log('>>>', component.children().render());
        // console.log('>>>', mockPush.mock);
      }
    });
  });

  it('should have cards that navigate to product IDs when clicked 2', () => {
    const mockPush = jest.fn();
    const props = {};
    jest.mock('react-router-dom', () => ({
      ...jest.requireActual('react-router-dom'),
      useHistory: () => ({
        push: mockPush
      })
    }));

    mockWindowLocation({
      url: 'https://ci.foo.redhat.com/loremIpsum/dolorSit/',
      callback: () => {
        const component = mount(<ProductViewMissing {...props} />);
        const cardButton = component.find(Card).first().find(Button).first();
        cardButton.simulate('click');
      }
    });

    // console.log('>>>', component.children().render());
    console.log('>>>', mockPush.mock);
  });

  /*
  it('should have cards that navigate to products when clicked', async () => {
    const props = {};

    // const mockPush = val => console.log('HEY >>>>>>>>>>>>>', val);
    // const mockPush = jest.fn();
    // jest.spyOn(ReactRouterDom, useHistory).mockImplementation(() => ({ push: mockPush }));

    /*
    const mockPush = jest.fn(val => val);
    jest.mock('react-router-dom', () => ({
      ...jest.requireActual('react-router-dom'),
      useHistory: () => ({
        push: mockPush
      })
    }));
     * /
    // mockPush.mockReturnValue();
    // jest.mock('react-router-dom', () => ({ useHistory: { push: () => mockPush } }));
    // jest.mock('react-router-dom', () => jest.fn(() => jest.fn(() => ({ push: mockPush }))));
    // jest.mock('react-router-dom', () => ({ useHistory: { push: () => mockPush } }));
    // jest.mock('react-router-dom', () => ({ useHistory: jest.fn(() => ({ push: () => mockPush })) }));
    // jest.mock('react-router-dom', () => jest.fn(() => ({ useHistory: { push: () => mockPush } })));// ({ useHistory: { push: () => mockPush } }));
    // jest.mock('react-router-dom', () => ({ useHistory: () => ({ push: mockPush }) }));

    /*
    jest.mock('react-router-dom', () => ({
      useHistory: () => {
        const push = () => mockPush;
        return { push };
      }
    }));
    * /
    // jest.mock('react-router-dom', () => ({ useHistory: { push: () => mockPush } }));

    /*
    jest.mock('react-router-dom', () => {
      const Test = function () { // eslint-disable-line
        this.useHistory = () => this;
        this.push = () => mockPush;
      };
      return new Test();
    });
    * /

    await mockWindowLocation({
      url: 'https://ci.foo.redhat.com/insights/subscriptions/',
      callback: () => {
        const component = mount(
          <MemoryRouter initialEntries={['/']}>
            <ProductViewMissing {...props} />
          </MemoryRouter>
        );

        // component.find(Button).first().simulate('click');
        // const spy = jest.spyOn(component, 'onSubmitOptIn');
        // expect(component).toMatchSnapshot('non-connected');
        // component.find(Button).at(1).simulate('click');
        // component.find(Card).first().find(Button).first().simulate('click', 'lorem ipsum');
        // const mockPush = jest.spyOn(component.instance(), 'onClick');
        const cardButton = component.find(Card).first().find(Button).first();

        cardButton.simulate('click', 'lorem ipsum');
        // console.log(component.find(Button).at(1).html());
        // console.log(component.find(Card).first().find(Button).html());
        console.log(cardButton.html());
        console.log(mockPush.mock.calls);

        /*
        await loadHookComponent(() => {
          component.find(Card).first().find(Button).first().simulate('click');
          // console.log(component.find(Button).at(1).html());
          console.log(component.find(Card).first().find(Button).html());
          console.log(mockPush.mock.calls);
          expect(mockPush).toHaveBeenCalledTimes(2);
        });
        * /
      }
    });
  });
  */
});
