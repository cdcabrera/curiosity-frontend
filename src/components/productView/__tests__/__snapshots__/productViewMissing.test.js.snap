// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductViewMissing Component should redirect when there are limited product cards: redirect action 1`] = `
[
  [
    "openshift-container",
  ],
]
`;

exports[`ProductViewMissing Component should render a basic component: basic 1`] = `
<PageLayout
  className="curiosity-missing-view"
>
  <PageHeader
    productLabel="missing"
    t={[Function]}
  >
    t(curiosity-view.title, {"appName":"Subscriptions"})
  </PageHeader>
  <PageSection
    isFilled={true}
  >
    <Gallery
      hasGutter={true}
    >
      <Card
        isHoverable={true}
        key="missingViewCard-OpenShift Container Platform-generatedid-"
        onClick={[Function]}
      >
        <CardTitle>
          <Title
            headingLevel="h2"
            size="lg"
          >
            t(curiosity-view.title, {"appName":"Subscriptions","context":"OpenShift Container Platform"})
          </Title>
        </CardTitle>
        <CardBody
          className="curiosity-missing-view__card-description"
        >
          t(curiosity-view.description, {"appName":"Subscriptions","context":"OpenShift Container Platform"})
        </CardBody>
        <CardFooter>
          <Button
            icon={
              <ArrowRightIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
            }
            iconPosition="right"
            isInline={true}
            onClick={[Function]}
            variant="link"
          >
            Open
          </Button>
        </CardFooter>
      </Card>
      <Card
        isHoverable={true}
        key="missingViewCard-OpenShift-dedicated-metrics-generatedid-"
        onClick={[Function]}
      >
        <CardTitle>
          <Title
            headingLevel="h2"
            size="lg"
          >
            t(curiosity-view.title, {"appName":"Subscriptions","context":"OpenShift-dedicated-metrics"})
          </Title>
        </CardTitle>
        <CardBody
          className="curiosity-missing-view__card-description"
        >
          t(curiosity-view.description, {"appName":"Subscriptions","context":"OpenShift-dedicated-metrics"})
        </CardBody>
        <CardFooter>
          <Button
            icon={
              <ArrowRightIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
            }
            iconPosition="right"
            isInline={true}
            onClick={[Function]}
            variant="link"
          >
            Open
          </Button>
        </CardFooter>
      </Card>
      <Card
        isHoverable={true}
        key="missingViewCard-rhacs-generatedid-"
        onClick={[Function]}
      >
        <CardTitle>
          <Title
            headingLevel="h2"
            size="lg"
          >
            t(curiosity-view.title, {"appName":"Subscriptions","context":"rhacs"})
          </Title>
        </CardTitle>
        <CardBody
          className="curiosity-missing-view__card-description"
        >
          t(curiosity-view.description, {"appName":"Subscriptions","context":"rhacs"})
        </CardBody>
        <CardFooter>
          <Button
            icon={
              <ArrowRightIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
            }
            iconPosition="right"
            isInline={true}
            onClick={[Function]}
            variant="link"
          >
            Open
          </Button>
        </CardFooter>
      </Card>
      <Card
        isHoverable={true}
        key="missingViewCard-RHEL-generatedid-"
        onClick={[Function]}
      >
        <CardTitle>
          <Title
            headingLevel="h2"
            size="lg"
          >
            t(curiosity-view.title, {"appName":"Subscriptions","context":"RHEL"})
          </Title>
        </CardTitle>
        <CardBody
          className="curiosity-missing-view__card-description"
        >
          t(curiosity-view.description, {"appName":"Subscriptions","context":"RHEL"})
        </CardBody>
        <CardFooter>
          <Button
            icon={
              <ArrowRightIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
            }
            iconPosition="right"
            isInline={true}
            onClick={[Function]}
            variant="link"
          >
            Open
          </Button>
        </CardFooter>
      </Card>
      <Card
        isHoverable={true}
        key="missingViewCard-rhods-generatedid-"
        onClick={[Function]}
      >
        <CardTitle>
          <Title
            headingLevel="h2"
            size="lg"
          >
            t(curiosity-view.title, {"appName":"Subscriptions","context":"rhods"})
          </Title>
        </CardTitle>
        <CardBody
          className="curiosity-missing-view__card-description"
        >
          t(curiosity-view.description, {"appName":"Subscriptions","context":"rhods"})
        </CardBody>
        <CardFooter>
          <Button
            icon={
              <ArrowRightIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
            }
            iconPosition="right"
            isInline={true}
            onClick={[Function]}
            variant="link"
          >
            Open
          </Button>
        </CardFooter>
      </Card>
    </Gallery>
  </PageSection>
</PageLayout>
`;
